apiserver/crossmodel/crossmodel_test.go

```diff
@@ -27,133 +26,204 @@ func (s *crossmodelSuite) SetUpTest(c *gc.C) {
+	expectedOffers := params.AddApplicationOffers{
+		Offers: []params.AddApplicationOffer{{
+			ApplicationOffer: params.ApplicationOffer{
+				ApplicationURL:         "local:/u/me/one",
+				SourceModelTag:         "model-deadbeef-0bad-400d-8000-4b1d0d06f00d",
+				SourceLabel:            "controller",
+				ApplicationName:        "one",
+				ApplicationDescription: "A pretty popular blog engine",
+				Endpoints: []params.RemoteEndpoint{
+					{Name: "db",
+						Role:      "requirer",
+						Interface: "mysql",
+						Limit:     1,
+						Scope:     "global",
+					},
+				}}}, {
+			ApplicationOffer: params.ApplicationOffer{
+				ApplicationURL:         "local:/u/me/two",
+				SourceModelTag:         "model-deadbeef-0bad-400d-8000-4b1d0d06f00d",
+				SourceLabel:            "controller",
+				ApplicationName:        "two",
+				ApplicationDescription: "A pretty popular blog engine",
+				Endpoints: []params.RemoteEndpoint{
+					{Name: "db",
+						Role:      "requirer",
+						Interface: "mysql",
+						Limit:     1,
+						Scope:     "global",
+					},
+				}}}}}
+	s.applicationDirectory.addOffers = func(offers params.AddApplicationOffers) (params.ErrorResults, error) {
+		c.Assert(offers, jc.DeepEquals, expectedOffers)
+		result := params.ErrorResults{}
+		result.Results = make([]params.ErrorResult, len(offers.Offers))
+		return result, nil
 	}
 
 	errs, err := s.api.Offer(all)
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(errs.Results, gc.HasLen, len(all.Offers))
 	c.Assert(errs.Results[0].Error, gc.IsNil)
-	s.serviceBackend.CheckCallNames(c, addOfferBackendCall)
+	c.Assert(errs.Results[2].Error, gc.IsNil)
+	c.Assert(errs.Results[1].Error, gc.ErrorMatches, `getting offered application notthere: application "notthere" not found`)
+	s.applicationDirectory.CheckCallNames(c, addOffersBackendCall)
 }
 
 func (s *crossmodelSuite) TestOfferError(c *gc.C) {
-	serviceName := "test"
-	s.addService(c, serviceName)
+	applicationName := "test"
+	s.addApplication(c, applicationName)
 	one := params.ApplicationOfferParams{
-		ApplicationName: serviceName,
+		ModelTag:        "model-deadbeef-0bad-400d-8000-4b1d0d06f00d",
+		ApplicationName: applicationName,
 	}
 	all := params.ApplicationOffersParams{[]params.ApplicationOfferParams{one}}
 
 	msg := "fail"
 
-	s.serviceBackend.addOffer = func(offer crossmodel.OfferedApplication, offerParams params.AddApplicationOffer) error {
-		return errors.New(msg)
+	s.applicationDirectory.addOffers = func(offers params.AddApplicationOffers) (params.ErrorResults, error) {
+		return params.ErrorResults{Results: []params.ErrorResult{{Error: &params.Error{Message: msg}}}}, nil
 	}
 
 	errs, err := s.api.Offer(all)
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(errs.Results, gc.HasLen, len(all.Offers))
 	c.Assert(errs.Results[0].Error, gc.ErrorMatches, fmt.Sprintf(".*%v.*", msg))
-	s.serviceBackend.CheckCallNames(c, addOfferBackendCall)
+	s.applicationDirectory.CheckCallNames(c, addOffersBackendCall)
 }
```

Comment:
Is there a test which covers some offers failing due to an offer params creation error?